Mathlib/SetTheory/Nimber/Field.lean

```diff
@@ -218,4 +217,127 @@ instance : CommRing Nimber where
 instance : IsDomain Nimber where
 instance : CancelMonoidWithZero Nimber where
 
+/-! ### Nimber division -/
+
+mutual
+
+/-- The nimber inverse `a⁻¹` is recursively defined as the smallest nimber not in the set `s`, which
+itself is recursively defined as the smallest set with `0 ∈ s` and `(1 + (a + a') * b) / a' ∈ s`
+for `0 < a' < a` and `b ∈ s`.
+
+This preliminary definition "accidentally" satisfies `invAux 0 = 1`, which the real inverse
+corrects. -/
+def invAux (a : Nimber) : Nimber :=
+  sInf (invSet a)ᶜ
+termination_by (a, 1)
+
+/-- The set in the definition of `invAux a`. -/
+def invSet (a : Nimber) : Set Nimber :=
+  ⋂₀ {s | 0 ∈ s ∧ ∀ a' < a, a' ≠ 0 → ∀ b ∈ s, invAux a' * (1 + (a + a') * b) ∈ s}
+termination_by (a, 0)
+
+end
+
+theorem zero_mem_invSet (a : Nimber) : 0 ∈ invSet a := by
+  rw [invSet]
+  exact Set.mem_sInter.2 fun _ hs => hs.1
+
+/-- "cons" is our operation `(1 + (a + a') * b) / a'` in the definition of the inverse. -/
+theorem cons_mem_invSet {a' : Nimber} (ha₀ : a' ≠ 0) (ha : a' < a) (hb : b ∈ invSet a) :
+    invAux a' * (1 + (a + a') * b) ∈ invSet a := by
+  rw [invSet] at hb ⊢
+  exact Set.mem_sInter.2 fun _ hs => hs.2 _ ha ha₀ _ (Set.mem_sInter.1 hb _ hs)
+
+/-- A recursion principle for `invSet`. -/
+@[elab_as_elim]
+theorem invSet_recOn {p : Nimber → Prop} (a : Nimber) (h0 : p 0)
+    (hi : ∀ a' < a, a' ≠ 0 → ∀ b, p b → p (invAux a' * (1 + (a + a') * b))) (x : Nimber)
+    (hx : x ∈ invSet a) : p x := by
+  revert x
+  rw [invSet]
+  exact Set.sInter_subset_of_mem ⟨h0, hi⟩
+
+/--- An auxiliary type for enumerating the elements of `invSet`, and proving that its complement
+is nonempty. -/
+private inductive InvTy (a : Nimber.{u}) : Type u
+  | zero : InvTy a
+  | cons : (toOrdinal a).toType → InvTy a → InvTy a
```

Comment:
isn't this isomorphic to `List (toOrdinal a).toType`?

Reply:
Nice catch! I guess my usage of `cons` should have given that away...